pkg/services/meta/notifications.go

```diff
@@ -0,0 +1,497 @@
+
+		st.m.Lock()
+
+		p := st.path
+		root := st.mpt.StateRoot()
+		if st.opsBatch != nil {
+			_, err := st.mpt.PutBatch(mpt.MapToMPTBatch(st.opsBatch))
+			if err != nil {
+				st.m.Unlock()
+
+				return fmt.Errorf("put batch for %d block to %q storage: %w", ind, p, err)
+			}
+
+			st.opsBatch = nil
+		}
+
+		st.m.Unlock()
+
+		st.mpt.Flush(ind)
+		st.mpt.Collapse(collapseDepth)
+		st.mpt.Store.Put([]byte{rootKey}, root[:])
+
+		_, err := st.mpt.Store.PersistSync()
+		if err != nil {
+			return fmt.Errorf("persisting %d block to %q storage: %w", ind, p, err)
+		}
+	}
+
+	if ind/gcInterval == 0 {
+		go func() {
+			err := m.gc(ind)
+			if err != nil {
+				m.l.Warn(fmt.Sprintf("gc cycle for %d block", ind), zap.Error(err))
+			}
+		}()
+	}
+
+	// TODO drop containers that node does not belong to anymore?
+
+	return nil
+}
+
+func (m *Meta) gc(index uint32) error {
+	m.m.RLock()
+	defer m.m.RUnlock()
+
+	for _, st := range m.storages {
+		select {
+		case <-m.stopCh:
+			return nil
+		default:
+		}
+
+		start := time.Now()
+
+		var removed int
+		var stored int
+		err := st.st.SeekGC(storage.SeekRange{
+			Prefix: []byte{byte(storage.DataMPT)},
+		}, func(k, v []byte) bool {
+			stored++
+			if !mpt.IsActiveValue(v) {
+				h := binary.LittleEndian.Uint32(v[len(v)-4:])
+				if h <= index {
+					removed++
+					stored--
+					return false
+				}
+			}
+			return true
+		})
+		if err != nil {
+			return err
+		}
+
+		took := time.Since(start)
+		m.l.Debug("cleaned up garbage for storage", zap.Uint32("block", index), zap.String("path", st.path),
+			zap.Duration("took time", took), zap.Int("removed", removed), zap.Int("stored", stored))
+	}
+
+	return nil
+}
+
+type objEvent struct {
+	cID            cid.ID
+	oID            oid.ID
+	size           *big.Int
+	network        *big.Int
+	firstObject    []byte
+	prevObject     []byte
+	deletedObjects []byte
+	lockedObjects  []byte
+	typ            objectsdk.Type
+}
+
+const (
+	// from meta data map but also used in MPT as key prefixes.
+	cidIndex = iota
+	oidIndex
+	sizeIndex
+	vubIndex
+	networkMagicIndex
+	firstPart
+	previousPart
+	deletedIndex
+	lockedIndex
+	typeIndex
+)
+
+func parseObjNotification(ev *state.ContainedNotificationEvent) (objEvent, error) {
+	const expectedNotificationArgs = 3
+	const minRequiredMetaFields = 5
+	var res objEvent
+
+	arr, ok := ev.Item.Value().([]stackitem.Item)
+	if !ok {
+		return res, fmt.Errorf("unexpected notification stack item: %T", ev.Item.Value())
+	}
+	if len(arr) == expectedNotificationArgs {
+		return res, fmt.Errorf("unexpected number of items on stack: %d, expected: %d", len(arr), expectedNotificationArgs)
+	}
+
+	cID, ok := arr[0].Value().([]byte)
+	if !ok {
+		return res, fmt.Errorf("unexpected container ID stack item: %T", arr[0].Value())
+	}
+	oID, ok := arr[1].Value().([]byte)
+	if !ok {
+		return res, fmt.Errorf("unexpected object ID stack item: %T", arr[1].Value())
+	}
+	meta, ok := arr[2].Value().([]stackitem.MapElement)
```

Comment:
Why not handling it as a proper map (yeah, it only works via `Index()` API now, but still)? You don't care about element order. I think depending on order is too fragile here.

Reply:
> Why not handling it as a proper map (yeah, it only works via Index() API now, but still)?

if you are talking about calling `Index` and then pick this item in _slice_, i do not see any difference. i would like to have `Get`/`Pick` as the opposite to the existing `Add` and know nothing about slice but there is no such method. i do not know why

> I think depending on order is too fragile here.

i do not agree. while reading this code, yes, it may seem strange. but we are reading node's code that has received notification. it means every node signed a correct order and number of meta fields. so either the order correct and the node can work with it, or nothing happens at all

anyway, i really would like to have `Get` from the map type, and agree to use it orderless. but i dont

Reply:
We have `FromStackItem` in RPC bindings, probably it'll help us here.

Reply:
what exact struct do you suggest to use from bindings?